roles/tests-integration/tests/pool_integration.rs

```diff
@@ -0,0 +1,22 @@
+mod common;
+
+use const_sv2::{
+    MESSAGE_TYPE_COINBASE_OUTPUT_DATA_SIZE, MESSAGE_TYPE_NEW_TEMPLATE,
+    MESSAGE_TYPE_SETUP_CONNECTION, MESSAGE_TYPE_SETUP_CONNECTION_SUCCESS,
+    MESSAGE_TYPE_SET_NEW_PREV_HASH,
+};
+
+#[tokio::test]
+async fn success_pool_template_provider_connection() {
+    let sniffer_addr = common::get_available_address();
+    let tp_addr = common::get_available_address();
+    let pool_addr = common::get_available_address();
+    let _tp = common::start_template_provider(tp_addr.port()).await;
+    let sniffer = common::start_sniffer(tp_addr, sniffer_addr).await;
+    let _ = common::start_pool(Some(pool_addr), Some(sniffer_addr)).await;
+    assert!(sniffer.expect_downstream_message(MESSAGE_TYPE_SETUP_CONNECTION));
+    assert!(sniffer.expect_upstream_message(MESSAGE_TYPE_SETUP_CONNECTION_SUCCESS));
+    assert!(sniffer.expect_downstream_message(MESSAGE_TYPE_COINBASE_OUTPUT_DATA_SIZE));
+    assert!(sniffer.expect_upstream_message(MESSAGE_TYPE_NEW_TEMPLATE));
+    assert!(sniffer.expect_upstream_message(MESSAGE_TYPE_SET_NEW_PREV_HASH));
+}
```

Comment:
let's also use `next_downstream_message` and `next_upstream_message` here so we can get a feeling of how the user experience is going to be like

Reply:
Maybe it is worth doing that in a test where it is needed and not here?

Reply:
ok here's a very simple test where I think this makes sense.

I couldn't find it as a MG test but honestly we should have had this in our CI for a long time.

- start TP
- start Pool (connected to TP)
- start SV2 CPU Miner (connected to Pool)

we can assert that:
- Pool connects to TP and receives template (basically what's already being done on `success_pool_template_provider_connection`)
- SV2 CPU Miner connects to Pool (`SetupConnection` / `SetupConnection.Success`)
- SV2 CPU Miner opens standard channel with Pool (`OpenStandardMiningChannel` / `OpenStandardMiningChannel.Success`)
- Pool sends `SetTarget` to SV2 CPU Miner (`SetTarget`)
- Pool sends Standard Job to SV2 CPU Miner (`NewMiningJob`)
- SV2 CPU Miner sends share to Pool (`SubmitSharesStandard`)
- Pool confirms receiving share (`SubmitShares.Success`)

we can assert things like:channel ids and job ids

Reply:
So, I was able to test the above but I think it requires some discussion/changes in `mining-device` crate. Currently `mining-device` calculates the difficulty in a way that does not always work. For me, the difficulty target was never hit when trying to send shares, even after running the test for 60 seconds. When lowering the difficulty target manually the mining-device sends the `SubmitShare` message succesfully, but it send many of them, and it will probably behave differently on different machines. I don't have a good solution for this yet, need to sleep on it.

To try to unblock this pr, made the following changes https://github.com/stratum-mining/stratum/pull/1066/commits/338cc02466a2e6b103e3a7cd84747b522c9f2828#diff-2a0b80af72100f3ef6475d64910e57bfa2194a668b3a7692391a53f964340fb4R21 to give an example of how to use `next_x_message`. Also made small doc/var name changes here https://github.com/stratum-mining/stratum/pull/1066/commits/e99e36b769a4b6a8a5ef39b1d86802b4eb95cfd1, will fixup this one once it is reviewed.